packages/modules/notification/src/services/notification-module-service.ts

```diff
@@ -0,0 +1,135 @@
+  protected baseRepository_: DAL.RepositoryService
+  protected readonly notificationService_: ModulesSdkTypes.InternalModuleService<TEntity>
+  protected readonly notificationProviderService_: NotificationProviderService
+
+  constructor(
+    {
+      baseRepository,
+      notificationModelService,
+      notificationProviderService,
+    }: InjectedDependencies,
+    protected readonly moduleDeclaration: InternalModuleDeclaration
+  ) {
+    // @ts-ignore
+    super(...arguments)
+    this.baseRepository_ = baseRepository
+    this.notificationService_ = notificationModelService
+    this.notificationProviderService_ = notificationProviderService
+  }
+
+  __joinerConfig(): ModuleJoinerConfig {
+    return joinerConfig
+  }
+  create(
+    data: NotificationTypes.CreateNotificationDTO[],
+    sharedContext?: Context
+  ): Promise<NotificationTypes.NotificationDTO[]>
+  create(
+    data: NotificationTypes.CreateNotificationDTO,
+    sharedContext?: Context
+  ): Promise<NotificationTypes.NotificationDTO>
+
+  @InjectManager("baseRepository_")
+  async create(
+    data:
+      | NotificationTypes.CreateNotificationDTO
+      | NotificationTypes.CreateNotificationDTO[],
+    @MedusaContext() sharedContext: Context = {}
+  ): Promise<
+    NotificationTypes.NotificationDTO | NotificationTypes.NotificationDTO[]
+  > {
+    const normalized = Array.isArray(data) ? data : [data]
+
+    const createdNotifications = await this.create_(normalized, sharedContext)
+
+    const serialized = await this.baseRepository_.serialize<
+      NotificationTypes.NotificationDTO[]
+    >(createdNotifications, {
+      populate: true,
+    })
+
+    return Array.isArray(data) ? serialized : serialized[0]
+  }
+
+  @InjectTransactionManager("baseRepository_")
+  protected async create_(
+    data: NotificationTypes.CreateNotificationDTO[],
+    @MedusaContext() sharedContext: Context = {}
+  ): Promise<TEntity[]> {
+    if (!data.length) {
+      return []
+    }
+
+    const notificationsToCreate = await promiseAll(
+      data.map(async (entry) => {
+        const provider =
+          await this.notificationProviderService_.getProviderForChannel(
+            entry.channel
+          )
+
+        if (!provider) {
+          throw new MedusaError(
+            MedusaError.Types.NOT_FOUND,
+            `Could not find a notification provider for channel: ${entry.channel}`
+          )
+        }
+
+        const res = await this.notificationProviderService_.send(
+          provider,
+          entry
+        )
+        return { ...entry, provider_id: provider.id, external_id: res.id }
+      })
+    )
+
+    // Currently we store notifications after they are sent, which might result in a notification being sent that is not registered in the database.
+    // If necessary, we can switch to a two-step process where we first create the notification, send it, and update it after it being sent.
+    const createdNotifications = await this.notificationService_.create(
```

Comment:
I believe we could save the notification before sending it. And after calling the provider we update its status.
This will have to be broken down into 3 different steps, because at the moment we are holding a DB transaction open while calling a 3rd party provider to send the notification.